File: ConstructBST.java
// Time Complexity :O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes, I was getting stack overflow error on line 48 
//which I could not figure out why

// Your code here along with comments explaining your approach:
//We are given postorder array which is left right root so pick the root element from the last index
//and recursively build the tree. Inorder array is being mapped to a HashMap so we do not require 
//it to be passed to helper function.

import java.util.HashMap;

 class TreeNode {
         int val;
         TreeNode left;
         TreeNode right;
         TreeNode() {}
         TreeNode(int val) { this.val = val; }
         TreeNode(int val, TreeNode left, TreeNode right) {
             this.val = val;
             this.left = left;
             this.right = right;
         }
     }

public class ConstructBST {
    private static HashMap<Integer,Integer> map = new HashMap<>();
    private static int index;
    public static TreeNode buildTree(int[] inorder, int[] postorder) {
        index = inorder.length - 1;
        if(postorder.length == 0)
            return null;
        for(int i = 0;i < inorder.length;i++)
        {
            map.put(inorder[i],i);
        }
        
        return helper(postorder,0,inorder.length - 1);
    }
    
    private static TreeNode helper(int[] postorder,int start,int end)
    {        
        //base
        if(start > end) //this means that the start and end have crossed each other
            return null;
        
        int rootVal = postorder[index];
        index--;
        int rootIndex = map.get(rootVal);
        TreeNode root = new TreeNode(rootVal);
        root.right = helper(postorder,rootIndex + 1,end);
        root.left = helper(postorder,start,rootIndex - 1);
        return root;
    }

    public static void main(String[] args) {
        int[] postorder = new int[] { 9, 15, 7, 20, 3 };
        int[] inorder = new int[] { 9, 3, 15, 20, 7 };

        TreeNode result = buildTree(inorder, postorder);
        System.out.println(result.val);
        System.out.println(result.left.val);
        System.out.println(result.right.val);
    }
}

File: construct_BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        :time complexity: O(n^2)
        :space complexity: O(n^2)
        """
        # # idea : zeroth index of pre order gives root node : root-->left-->right
        # # in order --> left-->root-->right
        if len(preorder)==0: return None
        root_index = 0
        root_val = preorder[0]
        root = TreeNode(root_val)
        for i in range(len(inorder)):
            if root_val == inorder[i]:
                root_index=i
        # Build four sub lists         
        pre_left = preorder[1:root_index+1]
        pre_right=preorder[root_index+1:len(preorder)]
        in_left=inorder[0:root_index]
        in_right=inorder[root_index+1:len(inorder)]
        root.left = self.buildTree(pre_left, in_left)
        root.right = self.buildTree(pre_right, in_right)
        return root
        
#         # shorter version
#         if not preorder or not inorder:
#             return None
#         root = TreeNode(preorder[0])
#         mid = inorder.index(preorder[0])
#         root.left = self.buildTree(preorder[1:mid+1], inorder[:mid])
#         root.right = self.buildTree(preorder[mid+1:], inorder[mid+1:])
#         return root
    
    